src/whitenoise/database/aggregated_messages.rs

```diff
@@ -560,6 +560,89 @@ impl AggregatedMessage {
             .collect())
     }
 
+    /// Fetches the most recent kind-9 message for each group in a single query.
+    ///
+    /// Returns `ChatMessageSummary` with `author_display_name: None`.
+    /// The caller populates display names after a separate user batch lookup.
+    ///
+    /// Groups without messages or with only deleted messages are not included
+    /// in the result.
+    pub async fn find_last_by_group_ids(
+        group_ids: &[GroupId],
+        database: &Database,
+    ) -> Result<Vec<ChatMessageSummary>> {
+        use sqlx::Row;
+
+        if group_ids.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        let group_id_bytes: Vec<Vec<u8>> = group_ids.iter().map(|id| id.to_vec()).collect();
+
+        // Build dynamic query with correct number of placeholders
+        let placeholders = "?,".repeat(group_id_bytes.len());
+        let placeholders = placeholders.trim_end_matches(',');
+
+        // Correlated subquery to get the last message per group
+        // Uses id matching with ORDER BY LIMIT 1 for deterministic results on timestamp ties
+        let query = format!(
+            "SELECT mls_group_id, author, content, created_at,
+                    json_array_length(media_attachments) as media_count
+             FROM aggregated_messages am1
+             WHERE kind = 9
+               AND mls_group_id IN ({})
+               AND deletion_event_id IS NULL
+               AND id = (
+                   SELECT id
+                   FROM aggregated_messages am2
+                   WHERE am2.mls_group_id = am1.mls_group_id
+                     AND am2.kind = 9
+                     AND am2.deletion_event_id IS NULL
+                   ORDER BY created_at DESC, id DESC
+                   LIMIT 1
+               )",
+            placeholders
+        );
```

Comment:
The correlated subquery executes once per row in am1, causing causing $O(n^2)$ time complexity, this is (mostly) fine because it doesn't get called that often, but if the user is part of many chats or they message a lot, it may be slower than it needs to be.
Why not use a CTE instead?

Reply:
something like this (please double check the query, did not test):

```sql
WITH ranked_messages AS (
    SELECT 
        mls_group_id,
        author,
        content,
        created_at,
        json_array_length(media_attachments) as media_count,
        ROW_NUMBER() OVER (
            PARTITION BY mls_group_id 
            ORDER BY created_at DESC, id DESC
        ) as rn
    FROM aggregated_messages
    WHERE kind = 9
      AND mls_group_id IN ({})
      AND deletion_event_id IS NULL
)
SELECT mls_group_id, author, content, created_at, media_count
FROM ranked_messages
WHERE rn = 1;
```